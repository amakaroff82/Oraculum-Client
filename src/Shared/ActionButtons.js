import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from 'material-ui/styles';
import Button from 'material-ui/Button';

const styles = theme => ({
  button: {
    margin: theme.spacing.unit,
  },
});

export const ConditionalButton = ({
  buttonProps: { textId, textValues, ...buttonProps },
  condition,
  ...rest
}) =>
  (condition === undefined || condition) && buttonProps ? (
    <Button {...buttonProps} {...rest}>
      "text val"
    </Button>
  ) : null;

ConditionalButton.propTypes = {
  buttonProps: PropTypes.shape({
    textId: PropTypes.string,
  }),
  condition: PropTypes.bool,
};

ConditionalButton.displayName = 'ConditionalButton';

export const ActionButtons = ({
  classes,
  primaryButtonProps: {
    condition: primaryButtonCondition,
    ...primaryButtonProps
  } = { condition: false },
  cancelButtonProps: {
    condition: cancelButtonCondition,
    ...cancelButtonProps
  } = { condition: false },
  extraButtons,
  children,
}) => (
  <div>
    <ConditionalButton
      raised
      className={classes.button}
      buttonProps={{ textId: 'cancel', ...cancelButtonProps }}
      condition={cancelButtonCondition}
    />
    {extraButtons &&
      extraButtons.map(({ condition, ...buttonProps }, index) => (
        <ConditionalButton
          key={index}
          raised
          className={classes.button}
          buttonProps={buttonProps}
          condition={condition}
        />
      ))}
    <ConditionalButton
      raised
      className={classes.button}
      buttonProps={primaryButtonProps}
      condition={primaryButtonCondition}
      color="primary"
    />
    {children}
  </div>
);

const ButtonPropsPropType = PropTypes.shape({
  condition: PropTypes.bool,
});

ActionButtons.propTypes = {
  classes: PropTypes.objectOf(PropTypes.string),
  primaryButtonProps: ButtonPropsPropType,
  cancelButtonProps: ButtonPropsPropType,
  extraButtons: PropTypes.arrayOf(ButtonPropsPropType),
  children: PropTypes.node,
};

ActionButtons.displayName = 'ActionButtons';

export default withStyles(styles, { name: 'ActionButtons' })(ActionButtons);
