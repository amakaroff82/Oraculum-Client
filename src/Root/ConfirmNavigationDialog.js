import React from 'react';
import Dialog, { DialogActions, DialogTitle } from 'material-ui/Dialog';
import ActionButtons from '../Shared/ActionButtons';

const ConfirmNavigationDialog = ({ confirmNavigation, updateNavDialog }) => {
  const handleCloseWithTransition = allowTransition => {
    updateNavDialog({ open: false });
    confirmNavigation.callback(allowTransition);
  };

  const handleCancel = () => handleCloseWithTransition(false);
  const handleConfirm = () => handleCloseWithTransition(true);

  return (
    <Dialog open={confirmNavigation.open} onRequestClose={handleCancel}>
      <DialogTitle>{confirmNavigation.message}</DialogTitle>
      <DialogActions>
        <ActionButtons
          primaryButtonProps={{ textId: 'confirm', onClick: handleConfirm }}
          cancelButtonProps={{ onClick: handleCancel }}
        />
      </DialogActions>
    </Dialog>
  );
};

export default ConfirmNavigationDialog;
