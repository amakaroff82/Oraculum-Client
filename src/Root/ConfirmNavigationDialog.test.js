import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import ConfirmNavigationDialog from './ConfirmNavigationDialog';
import ActionButtons from '../Shared/ActionButtons';

const defaultProps = {
  confirmNavigation: {
    open: true,
    message: 'Hello World',
    callback: jest.fn(),
  },
  updateNavDialog: jest.fn(),
};

const setup = props => {
  props = { ...defaultProps, ...props };

  return shallow(<ConfirmNavigationDialog {...props} />);
};

describe('<ConfirmNavigationDialog />', () => {
  let tree;

  beforeAll(() => {
    tree = setup();
  });

  it('renders snapshot correctly with required props', () => {
    expect(toJson(tree)).toMatchSnapshot();
  });

  describe('performs correct behavior on cancel button click', () => {
    beforeAll(() => {
      tree
        .find(ActionButtons)
        .props()
        .cancelButtonProps.onClick();
    });

    afterAll(() => {
      defaultProps.confirmNavigation.callback.mockReset();
      defaultProps.updateNavDialog.mockReset();
    });

    it('closes the confirmation dialog', () => {
      expect(defaultProps.updateNavDialog).toBeCalledWith({ open: false });
    });

    it('calls the confirm navigation callback with correct arg', () => {
      expect(defaultProps.confirmNavigation.callback).toBeCalledWith(false);
    });
  });

  describe('performs correct behavior on confirm button click', () => {
    beforeAll(() => {
      tree
        .find(ActionButtons)
        .props()
        .primaryButtonProps.onClick();
    });

    afterAll(() => {
      defaultProps.confirmNavigation.callback.mockReset();
      defaultProps.updateNavDialog.mockReset();
    });

    it('closes the confirmation dialog', () => {
      expect(defaultProps.updateNavDialog).toBeCalledWith({ open: false });
    });

    it('calls the confirm navigation callback with correct arg', () => {
      expect(defaultProps.confirmNavigation.callback).toBeCalledWith(true);
    });
  });

  describe('performs correct behavior on backdrop click', () => {
    beforeAll(() => {
      const backDropClickHandler = tree.props().onRequestClose;
      backDropClickHandler();
    });

    it('closes the confirmation dialog', () => {
      expect(defaultProps.updateNavDialog).toBeCalledWith({ open: false });
    });

    it('calls the confirm navigation callback with correct arg', () => {
      expect(defaultProps.confirmNavigation.callback).toBeCalledWith(false);
    });
  });
});
