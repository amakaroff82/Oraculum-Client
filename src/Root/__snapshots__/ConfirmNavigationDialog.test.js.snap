// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConfirmNavigationDialog /> renders snapshot correctly with required props 1`] = `
<withStyles(Dialog)
  onRequestClose={[Function]}
  open={true}
>
  <withStyles(DialogTitle)>
    Hello World
  </withStyles(DialogTitle)>
  <withStyles(DialogActions)>
    <withStyles(ActionButtons)
      cancelButtonProps={
        Object {
          "onClick": [Function],
        }
      }
      primaryButtonProps={
        Object {
          "onClick": [Function],
          "textId": "confirm",
        }
      }
    />
  </withStyles(DialogActions)>
</withStyles(Dialog)>
`;
